### ISSUE TITLE
adding alias for some commands

### ISSUE THREAD

**Issue by hainm** [Amber-MD/cpptraj#179]
**updated** I am adding a list of commands could be aliased. The word will be ~~crossed~~ if the idea is not accepted.

- ~~[ ] `hieragglo` with `hclust`~~
I swear that I always get trouble pronouncing 'hieragglo'. What's about make an alias `hclust`?

This key word is main stream in R world

https://stat.ethz.ch/R-manual/R-devel/library/stats/html/hclust.html

- [ ] center of mass

**Comment by hainm** [comment]
```python
cluster hclust
```

vs

```python
cluster hieragglo
```

**Comment by swails** [comment]
I think it's too easy to mistake `hclust` as something to do with Hydrogen (e.g., `hbond`).  `hieragglo` at least indicates the hierarchical agglomerative algorithm in an unambiguous way.

**Comment by hainm** [comment]
thanks for you comment. I am changing the title so I can add suggestion about word alias for other cases instead of creating new issue)

**Comment by drroe** [comment]
> What's about make an alias hclust?

I don't like it for two reasons. First, it's not that much of a savings to type `hclust` over `hieragglo`. Second, the keywords in cpptraj are already a bit too obscure, and `hclust` is way less information than `hieragglo`. For example, in the context of clustering algorithms `hclust` could mean "hierarchical agglomerative", "hierarchical divisive", "HCS clustering", etc.

Also, since it's the default algorithm, just don't specify any algorithm for your clustering command and you won't have to remember `hieragglo` :-)

**Comment by swails** [comment]
I always start -1 on aliases and have to be convinced they're good (guilty until proven innocent).  Preserving backwards compatibility and supporting naturally synonymous keywords (like `sulphur` being an alias for `sulfur` or `aluminium` for `aluminum`) are examples of convincing aliases for me.

IMO, saving typing is rarely a good excuse when you're talking about a difference of less than a dozen characters (e.g., `hieragglo` vs `hclust`).  I'd get behind making `hieragglo` an alias for `hierarchicalagglomerativeclustering` because the latter is obnoxious, but there's nothing like that in cpptraj right now.  Although this falls into the "keep it around to maintain backwards compatibility" category.

**Comment by drroe** [comment]
> I'd get behind making hieragglo an alias for hierarchicalagglomerativeclustering because the latter is obnoxious, 

Now that you mention it, the keyword should probably be `theAlgorithmIwantToUseIsTheHierarchicalAgglomerativeMethodForClustering`

Case sensitive.

**Comment by hainm** [comment]
I agree with Jason and Dan. But in some cases we still need to make alias (for backward compat?), for example:

`dihedralscan`: This cause confusion for user (http://archive.ambermd.org/201506/0061.html)

**Comment by drroe** [comment]
> This causes confusion to user (http://archive.ambermd.org/201506/0061.html)

OK, the `dihedralscan` command may not be well-named, but in my opinion you shouldn't *guess* at what a command does and then get annoyed when it doesn't do what you expect. There is a manual entry for `dihedralscan` that describes exactly what it does. 
```
This command can be used to either generate a trajectory in which all specified dihedrals are rotated by a specified
amount, or all specified dihedrals are rotated by a random amount. Note that ’interval’ with outtraj should only
be used with one frame, otherwise the resulting trajectories will be overwritten.
```
I think that adding command aliases should not be a decision that is made lightly. Some of them are very obvious and make sense to me: for example, using `rmsf` for `atomicfluct` and `dssp` for `secstruct`, because in those cases the "alias" was in fact as good or even better a description of what the command does than the original command. It's not as clear what would be a better name for `dihedralscan`. In that email thread I suggested `rotatedihedrals`, but there is in fact an issue open now (#36) that proposes implementing a `rotatedihedral` command as a simpler alternative to `makestructure`. 

The bottom line is that cpptraj contains a ton of commands now, so to avoid polluting the command space I want to try and be cautious about adding new things. Make sense?

**Comment by hainm** [comment]
> The bottom line is that cpptraj contains a ton of commands now, so to avoid polluting the command space I want to try and be cautious about adding new things. Make sense?

yeah, fair enough.

**Comment by swails** [comment]
There are 2 types of aliases being suggested here.  Command aliases and keyword aliases.  As I mentioned, I'm -1 on both by default, but can be convinced to go +0/1 on a case-by-case basis.

`secstruct`, `rmsf`, `dssp`, and `rmsd` are convincing aliases for me because they are just as descriptive, just as obvious about what they should do, and it would be horribly confusing if they did something different (so there's no chance of eliminating a name you might use somewhere else later).

Command aliases are *worse* in that they occupy a global namespace that may negatively affect future plans.

Keyword aliases are still *bad* in that they increase the complexity of each command (and the commands are already too complex).  But they are at least more localized.

> OK, the dihedralscan command may not be well-named, but in my opinion you shouldn't guess at what a command does and then get annoyed when it doesn't do what you expect. There is a manual entry for dihedralscan that describes exactly what it does.

I only partially agree here (bigger part agree than not, though).  With `dihedralscan`, I agree completely -- the command name is clearly not self-documenting.  There is no single command name that can uniquely specify what that command does (at least that I can think of).  Obvious thing to do is look it up in the manual or online help.

But some commands *should* be self-documenting -- like `rmsd`.  I reserve the right to be annoyed at any package that makes a `rmsd` command do something *besides* compute a root mean squared deviation :).

**Comment by drroe** [comment]
Remaining open issues seem addressed by #238.

**Comment by hainm** [comment]
should cpptraj has explicit command for center of mass? (vs `vector ...`)

**Comment by drroe** [comment]
> should cpptraj has explicit command for center of mass? (vs vector ...)

Do you mean the functionality covered by `vector center`? If so, why? If new functionality, could you describe what it would do in more detail?

**Comment by hainm** [comment]
uhm, actually just name alias. For example, new users are more familiar with the name 'center of mass' (or center_of_mass) rather 'vector center mass'. This is from my own experience with cpptraj, it took times for me to figure out how to compute center of mass in cpptraj.

**Comment by drroe** [comment]
> For example, new users are more familiar with the name 'center of mass' (or center_of_mass) rather 'vector center mass'.

This is creating an individual command alias for part of a command, which isn't sustainable since there is no way to ensure that 'center_of_mass' isn't then just used for other vector commands without some ugly code in Action_Vector itself that changes behavior based on the command name. The other option would be to split off the functionality into a separate Action, but the whole reason `vector center` is part of Action_Vector is shared functionality (also, that then breaks backwards compatibility). 

I think the correct way to address things which may be confusing for new users is to have tutorials for them.

**Comment by hainm** [comment]
> I think the correct way to address things which may be confusing for new users is to have tutorials for them.

:+1:

**Comment by drroe** [comment]
This seems good to close.
